text/3334-niche.md

```diff
@@ -0,0 +1,348 @@
+
+We could guarantee more usage of niches than just a single value; however, this
+would constrain the compiler in areas that still see active development.
+
+We could avoid guaranteeing the use of a single-value niche for `Option`;
+however, this would eliminate one of the primary user goals for such niches.
+
+We could require types to opt into the guaranteed use of a niche, separately
+from declaring a niche. This seems unnecessarily verbose, as well as limiting:
+we can't yet provide a full guarantee of all *future* uses we may want to
+guarantee, only of the limited single-value uses.
+
+We could implement niches using a lang-item type that uses const generics (e.g.
+`Niche<T, const RANGE: std::ops::Range<T>>`. This type would be useful
+regardless, and we should likely provide it if we can. However, this RFC
+advocates (eventually) building such a type on an underlying language-level
+building block like `niche`, and providing the underlying building blocks to
+the ecosystem as well.
+
+We could implement niches using a trait `Niche` implemented for a type, with
+associated consts for invalid values. If we chose to do this in the future, the
+`#[niche(...)]` attribute could become forward-compatible with this, by
+generating the trait impl.
+
+We could use a syntax based on patterns, such as `struct S(u8 is 0..=32);` or
+`struct S(MyEnum is MyEnum::A | MyEnum::B)`.
+
+# Prior art
+[prior-art]: #prior-art
+
+The Rust compiler has supported niches for types like `Option` in various forms
+since versions prior to Rust 1.0. In particular, Rust 1.0 already guaranteed
+that `Option<&T>` has the same size as `&T`. Rust has had many additional
+niche-related optimizations since then.
+
+The Rust compiler already supports user-defined niches via the unstable
+attributes `rustc_layout_scalar_valid_range_start` and
+`rustc_layout_scalar_valid_range_end`.
+
+Bit-twiddling tricks to store information compactly have seen widespread use
+and innovation since computing antiquity.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+Does the compiler support niches on structs containing ZST fields such as
+`PhantomData`? If it doesn't, then initially, having a limitation to only
+structs containing a single field would be fine, and would not substantially
+reduce the usefulness of stabilizing this feature.
+
+Could we support niches on generic types? For instance, could we support
+declaring a niche of `0` on a generic structure with a single field?
+
+Could we support negative numbers in a niche attribute, at least for fields of
+concrete primitive type? That would provide a much more friendly interface, but
+would require the compiler to better understand the type and its size.
+
+Will something go wrong if applying a niche to a struct whose non-ZST field is
+itself a struct containing multiple fields? Do we need to restrict niches to
+structs containing primitive types, or similar?
```

Comment:
I think if non-primitive fields are supported, this RFC needs to do a lot more specifying:

* Niche-field-types without a guaranteed representation (e.g. structs without `#[repr(C)]` or `#[repr(transparent)]`, not sure about enum layout) should produce a warning or error.
* What if the niche-field-type itself already has niches?
* How does the niche interact with padding?
* If non-primitive fields are supported, endianness becomes relevant:
  
  ```rust
  #[niche(value = 0xFF)]
  struct S { v: [u8; 2] }
  ```
  
   Is the niche `[0xFF, 0x00]` or `[0x00, 0xFF]`?
   
  (Actually, explicitly specifying interaction with endianness in general may be sensible, irrespective of whether non-primitive niches are supported or not).

Reply:
@TimNN Yeah, that's a convincing argument. In that case, I think niches of non-primitive types will probably need to wait until we can specify values/ranges in terms of the field types.

I've now simplified the RFC further, requiring only simple field types.